packages/aws-cdk/lib/api/logs/logs-monitor.ts

```diff
@@ -0,0 +1,248 @@
+  /**
+   * The log event message
+   */
+  readonly message: string;
+
+  /**
+   * The name of the log group
+   */
+  readonly logGroup: string;
+
+  /**
+   * The time at which the event occurred
+   */
+  readonly timestamp: Date;
+}
+
+/**
+ * Options for adding log groups to the CloudWatchLogEventMonitor
+ */
+export interface AddLogGroupOptions {
+  /**
+   * The AWS account that the log group lives in
+   */
+  readonly account: string;
+
+  /**
+   * The AWS region that the log group lives in
+   */
+  readonly region: string;
+
+  /**
+   * The SDK for a given account
+   */
+  readonly sdk: ISDK;
+
+  /**
+   * A map of log groups and startTime in a given account
+   */
+  readonly groups: Set<string>;
+}
+
+/**
+ * Configuration tracking information on the log groups that are
+ * being monitored
+ */
+export interface LogGroupConfig {
+  /**
+   * The SDK for a given account
+   */
+  readonly sdk: ISDK;
+
+  /**
+   * A map of log groups and startTime in a given account
+   */
+  readonly groups: Map<string, number>;
+}
+
+/**
+ * Options for creating a CloudWatchLogEventMonitor
+ */
+export interface CloudWatchLogEventMonitorOptions {
+  /**
+   * The time that watch was first triggered.
+   * CloudWatch logs will be filtered using this start time
+   *
+   * @default - the current time
+   */
+  readonly hotswapTime?: Date;
+}
+
+export class CloudWatchLogEventMonitor {
+  /**
+   * Determines which events not to display
+   */
+  private startTime: number;
+
+  /**
+   * Map of account to LogGroupConfig
+   */
+  private readonly logGroups = new Map<string, LogGroupConfig>();
+
+  private active = false;
+
+  constructor(options: CloudWatchLogEventMonitorOptions = {}) {
+    this.startTime = options.hotswapTime?.getTime() ?? Date.now();
+    this.scheduleNextTick();
+  }
+
+  /**
+   * resume reading/printing events
+   */
+  public activate(): void {
+    this.active = true;
+  }
+
+  /**
+   * deactivates the monitor so no new events are read
+   * use case for this is when we are in the middle of performing a deployment
+   * and don't want to interweave all the logs together with the CFN
+   * deployment logs
+   *
+   * Also resets the start time to be when the new deployment was triggered
+   * and clears the list of tracked log groups
+   */
+  public deactivate(): void {
+    this.active = false;
+    this.startTime = Date.now();
+    this.logGroups.clear();
+  }
+
+  /**
+   * Adds CloudWatch log groups to read log events from.
+   * Since we could be watching multiple stacks that deploy to
+   * multiple environments (account+region), we need to store a list of log groups
+   * per env along with the SDK object that has access to read from
+   * that AWS account.
+   */
+  public addLogGroups(options: AddLogGroupOptions): void {
+    const env = `${options.account}:${options.region}`;
+    const groups = new Map<string, number>();
+    options.groups.forEach(group => {
+      groups.set(group, this.startTime);
+    });
```

Comment:
Can we use `reduce` here?

```suggestion
    const groupsStartTimes = options.groups.reduce((acc, groupName) => {
      acc[groupName] = this.startTime;
      return acc;
    }, {} as { [groupName: string]: number });
```